InteractiveUtils: Support callable objects as functions in introspection macros

serenity4 opened the pull request:
Follow-up to the follow-up https://github.com/JuliaLang/julia/pull/57911, building on the changes to introspection functions to support signature tuples being provided as a single argument.

This enables support for calls of the form
```julia
@code_typed (::Returns{Int})(1)
@code_llvm (::Base.Fix2{typeof(+), Float64})(::Int)
```
by providing an extra `use_signature_tuple::Bool = false` parameter in `gen_call_with_extracted_types`. Setting this parameter to true changes the code generation from `$fcn(f, Tuple{argtypes...})` to `$fcn(Tuple{f, argtypes...})` (where `$fcn` can be e.g. `code_typed`, `code_llvm` etc).

topolarity committed:
Support "Functor-like" `code_typed` invocation

This allows you to call, e.g., `code_typed((Foo, Int, Int))` to query
the code for `(::Foo)(::Int, ::Int)`

topolarity committed:
Support "Functor-like" `code_typed` invocation

This allows you to call, e.g., `code_typed((Foo, Int, Int))` to query
the code for `(::Foo)(::Int, ::Int)`

topolarity committed:
Base: Support functors in `method_instance(s)`, `code_ircode`, and `code_lowered`

topolarity committed:
Base: Support functors in `method_instance(s)`, `code_ircode`, and `code_lowered`

topolarity committed:
InteractiveUtils: Support functors in `code_llvm`, `code_native`, and `code_warntype`

topolarity committed:
InteractiveUtils: Support functors in `code_llvm`, `code_native`, and `code_warntype`

serenity4 committed:
Add support for directly providing signature tuple

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into interactive-utils-functors

serenity4 committed:
Merge branch 'ct/functor-reflection' of github.com:topolarity/julia into interactive-utils-functors

serenity4 committed:
Merge branch 'master' into interactive-utils-functors

serenity4 commented:
Oh, I think we had a branch on `isa(f, OpaqueClosure)` in the two-argument form of `code_typed` :thinking:

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into interactive-utils-functors

topolarity commented:
>  otherwise all packages relying on `gen_call_with_extracted_types_and_kwargs` will stop working.

Isn't that already an issue with switching to the 1-arg form from the 2-arg form?

topolarity commented:
Yeah, I agree with your breakdown.

I don't have a strong feeling among the alternatives, so I'd be fine if you want to pick one and see if the implementation feels all right.

serenity4 committed:
Improve error message

serenity4 committed:
Merge branch 'interactive-utils-functors' of github.com:serenity4/julia; branch 'master' of github.com:JuliaLang/julia into interactive-utils-functors

serenity4 committed:
Fix implementation for OpaqueClosure

serenity4 committed:
Add additional test

topolarity reviewed (approved):
Probably worth documenting the macro-dev-facing requirements for setting these boolean flags (I already forgot since last time, hah), but otherwise looks great.

serenity4 commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -398,13 +401,13 @@ end
               :infer_return_type, :infer_effects, :infer_exception_type]
     @eval macro ($fname)(ex0...)
-        gen_call_with_extracted_types_and_kwargs(__module__, $(QuoteNode(fname)), ex0; is_source_reflection = false)
+        gen_call_with_extracted_types_and_kwargs(__module__, $(QuoteNode(fname)), ex0; is_source_reflection = false, use_signature_tuple = $(in(fname, [:code_warntype, :code_llvm, :code_native])))
```
Indeed!

Yes, I forgot to enable it, thanks.

serenity4 committed:
Document `gen_call_with_extracted_types`

serenity4 committed:
Enable support for callable objects for source reflection macros

serenity4 commented:
@topolarity does the docstring look clear enough to you? I think I included pretty much everything that developers need to know. Suggestions welcome!

serenity4 committed:
Add `kws...` in docs for `fcn` usage

serenity4 committed:
Merge branch 'master' into interactive-utils-functors

serenity4 commented:
Yes, I started yesterday but https://github.com/JuliaLang/julia/pull/59239 brought a tough conflict (not a bad one though!). It changed the way that macro hygiene is handled in `gen_call_with_extracted_types`, and it took me some time to figure out what we should do. Notably because I also rewrote how we handle hygiene in this PR, and in a different direction than this other PR - and the design there is more composable than here so I'll most likely adopt it (unless I run into issues).

serenity4 committed:
Properly handle (re)escaping and replace_ref_begin_end! interactions

This commit refactors some of the recent additions to handle reescaping
in InteractiveUtils, along with tests to cover various hygiene contexts.

`replace_ref_begin_end!` was recently replaced in InteractiveUtils
by its low-level counterpart, `replace_ref_begin_end_!`, to better
handle escapes. However, we were using it in a way that
required duplication of the code generation logic to match the code
that we emit it for other code paths. Because the logic is now more complex with
the recently added `use_signature_tuple` behavior, it is preferable to
allow reuse of the same code path for all code we emit. This commit
implements this refactor, which unfortunately also requires a manual
escape of the temporary variable that is emitted by
`replace_ref_begin_end!` in its current design.

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into interactive-utils-functors

serenity4 committed:
Fix doctest failure

This replaces the example with a macro call
that has a simpler output, now that `edit`
generates a bigger expression (as it now
handles opaque closure arguments)

serenity4 commented:
@topolarity feel free to take a quick look at the recent commits, and otherwise merge. Thanks for having followed this PR until now :100:

topolarity reviewed (approved):
I haven't reviewed the hygiene changes, but otherwise this looks great to me.

Thanks for all the follow-through @serenity4

vtjnash commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -21,25 +22,33 @@ function rewrap_where(ex::Expr, where_params::Union{Nothing, Vector{Any}})
-    # Always unescape to get the core expression, then reescape and esc
-    original_ex = ex
-    ex = Meta.unescape(ex)
+function reescape(f::Function, @nospecialize ex)
+    isa(ex, Expr) || return f(ex)
+    unescaped = Meta.unescape(ex)
+    new = f(unescaped)
+    return reescape(new, ex)
+end
```
Why are you vendoring a buggy copy of Meta.reescape now?

serenity4 replied:
What is buggy with its implementation? My understanding is that the pattern is:
- Get the escaped argument (with `Meta.unescape`)
- Process it to possibly generate a new `Expr` (that's "user"-defined)
- Re-apply the original hygiene environment (escapes, hygienic scopes etc)

serenity4 replied:
The `reescape(f::Function, ex)` method is specifically to apply the pattern from #59239 where we [unescape to get the core expression, then reescape](https://github.com/JuliaLang/julia/blob/8190fe87b5dffd7dd1f15939aa7c14e5f7332be8/stdlib/InteractiveUtils/src/macros.jl#L107). I'd be happy to change it if it does something wrong, but the pattern it automates looks correct to me.

I agree that the other (copied) `reescape` method can be removed. I have a question though, why do we remove `:var"hygienic-scope"` with `unescape` but [don't reapply it](https://github.com/JuliaLang/julia/blob/cbea8cf28e0f617333130add1c4ab2a51f6c979c/base/meta.jl#L640) with `reescape` (and apply `escape` instead regardless if the original was an `:escape` or `:var"hygienic-scope"` expression)?

vtjnash commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -206,15 +211,167 @@ function merge_namedtuple_types(nt::Type{<:NamedTuple}, nts::Type{<:NamedTuple}.
+    assignment = decl.args[1]
+    isexpr(assignment, :(=), 2) || return
+    variable = assignment.args[1]
+    startswith(string(variable), "##S#") || return
```
This seems like a bad idea to find some random variable by name and then apply random mutation to it (mainly it seems it would result in incorrect code being executed if someone has a statement with multiple `begin` or `end` like `a[end][end]`)

vtjnash commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -292,41 +450,24 @@ function gen_call_with_extracted_types(__module__, fcn, ex0, kws = Expr[]; is_so
             nt = generate_merged_namedtuple_type(kwargs)
-            tt = rewrap_where(:(Tuple{$nt, $(get_typeof.(args)...)}), where_params)
-            return :($(fcn)(Core.kwcall, $tt; $(kws...)))
+            nt = Ref(nt) # ignore `get_typeof` handling
```
This seems very sketchy, since `Ref(nt)` is a perfectly fine, and somewhat common, thing to normally encounter in an AST already (after this function called macroexpand). There should not be any unique behavior of calling `typeof` based on the value of a object found in the code.

vtjnash commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -342,75 +483,46 @@ function gen_call_with_extracted_types(__module__, fcn, ex0, kws = Expr[]; is_so
-            return ex
         else
-            PairSymAny = Pair{Symbol, Any}
-            for (head, f) in (PairSymAny(:hcat, Base.hcat),
```
Please reapply this change to add PairSymAny, so that this part of the code can be type-stable

serenity4 replied:
Is there any actual benefit in doing so? I would mind in most cases, but not in

vtjnash replied:
I guess not, just seemed cleaner for future linters

serenity4 replied:
I think we have quite a few places (probably a lot of expression-processing functions notably) where we iterate `::Vector{Any}` anyways, instead of iterating over indices then index into the array (which I guess might be the "best" pattern for type-stability).

serenity4 commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -262,10 +420,10 @@ function gen_call_with_extracted_types(__module__, fcn, ex0, kws = Expr[]; is_so
             fully_qualified_symbol &= ex1 isa Symbol
             if fully_qualified_symbol || isexpr(ex1, :(::), 1)
-                call_reflection = :($(fcn)(Base.getproperty, $(typesof_expr(ex0.args, where_params))))
+                call_reflection = gen_call(fcn, [getproperty; ex0.args], where_params, kws; use_signature_tuple)
```
Ah, yeah I didn't notice this array was not typed with `Any[`. Performance/style-wise I thought that `;` was preferred for this pattern, but I don't mind at all using `...`.

vtjnash replied:
```
julia> a,b... = Any[1,2,3]
3-element Vector{Any}:
 1
 2
 3

julia> b
2-element Vector{Any}:
 2
 3
```

`...` implies `vect` and `;` implies `vcat`. It is thus strange to have vcat when some of the elements are not arrays, though perhaps efficient. With `Any[...]` it actually implies `getindex`, and `getindex(::Type{Any})` is marked no specialize.
